packages/@aws-cdk/cdk/lib/core/tokens.ts

```diff
@@ -133,3 +179,128 @@ export function resolve(obj: any, prefix?: string[]): any {
 
     return result;
 }
+
+import { FnConcat } from "../cloudformation/fn";
+
+/**
+ * Central place where we keep a mapping from Tokens to their String representation
+ *
+ * The string representation is used to embed token into strings,
+ * and stored to be able to
+ *
+ * All instances of TokenStringMap share the same storage, so that this process
+ * works even when different copies of the library are loaded.
+ */
+class TokenStringMap {
+    private readonly tokenMap: {[key: string]: Token};
+
+    constructor() {
+        const glob = global as any;
+        this.tokenMap = glob.__cdkTokenMap = glob.__cdkTokenMap || {};
+    }
+
+    /**
+     * Return a unique string for this Token
+     *
+     * Every call for the same Token will produce a new unique string, no
+     * attempt is made to deduplicate. Token objects should cache the
+     * value themselves, if required.
+     *
+     * If the token is an IRepresentableToken, it chooses its own representation
+     * string. This may be used to produce aesthetically pleasing and recognizable
+     * token representations for humans.
+     */
+    public register(token: Token | HintedToken): string {
+        const counter = Object.keys(this.tokenMap).length;
+        const representation = isHintedToken(token) ? token.representationHint : `TOKEN`;
+
+        const key = `${representation}.${counter}`;
+        if (new RegExp(`[^${VALID_KEY_CHARS}]`).exec(key)) {
+            throw new Error(`Invalid characters in token representation: ${key}`);
+        }
+
+        this.tokenMap[key] = token;
```

Comment:
Defensive: check that `key` doesn't already exist?

Reply:
Can't really happen since every one gets a unique number that's guaranteed to not occur in the map yet.

Reply:
I guess